### ISSUE TITLE
Incorrect completion for var in deconstruction

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#14901]
If you type space in the context below (with the cursor at the $$ mark), you get a completion to "VarAnnoying", when you really only mean to type "var".

```C#
class Program
{
    static void Main(string[] args)
    {
        (var$$ )
    }

    static (int count, double mean) Stats()
    {
        return (2, 1.5);
    }
}
class VarAnnoying { }
```

FYI @CyrusNajmabadi

**Comment by jcouv** [comment]
@CyrusNajmabadi As I was fixing this I ran into a related completion problem: when typing `(var a$$`, I get offered "as" and "is".
I'm thinking to remove comma and close-paren as completion triggers (when inside a possible tuple context). Does that sound right?

**Comment by CyrusNajmabadi** [comment]
Isn't this definitely a declaration location 100% of the time?  If so, why is completion coming up?  If not, when is this not a declaration?

**Comment by jcouv** [comment]
There is a valid case for offering "as", in an expression like `(x as Type)?.M();`.
`IsIsOrAsContext` has some heuristics for distinguishing when `x` is an expression vs. a type, which I'm trying to figure out.

**Comment by CyrusNajmabadi** [comment]
ah ok.

in that case, we should make this a place where it could be a reference/declaration.  i.e. we should have a builder here, and 'as' should not be preselected

**Comment by jcouv** [comment]
I'm not sure what you mean by "a builder".
More or less, I'm expecting to implement a `ShouldPreselect` method on that keyword recommender, and then the unittest should expect a soft-selection. Is that right?

**Comment by CyrusNajmabadi** [comment]
So, if you type something like:

```
Func<int> f = a
```

'a' is a location that could be a reference or a declaration.  i.e. you could be referencing something that exists with 'a' in it, but it could also be the declaration of a lambda parameter.

So we pop up a 'builder'.  An extra item that indicates that you could be 'building a new declaration name'.  So the rules are:

1. Always a reference: pop up a normal completion list.
2. Always a declaration: do not pop up the completion list.
3. Could be a reference or declaration: pop up the completion list, but have a builder so that completion doesn't interfere with your typing.

**Comment by jcouv** [comment]
Fixed in RC2 by #15175
